### PR TITLE
FEATURE: Introduce `NodeIdentity`

### PR DESCRIPTION
**Pull request by mhsdesign**
Resolves partially: https://github.com/neos/neos-development-collection/issues/4564
Required for a new `NodeUriBuilder` api https://github.com/neos/neos-development-collection/issues/4552 which in-turn "should" solve cross cr linking: https://github.com/neos/neos-development-collection/issues/4441

The `NodeIdentity` will contain `WorkspaceName` instead of the `ContentStreamId`

For most, if not all use cases, like caching a reference to a node in fusions @cache.context or serializing a node into the frontend and passing it around as query parameter, the workspace name should be used.
    
That guarantees that a node is still findable, when a workspace will be rebased (which _can_ happen anytime, and thus the csid would be changed

**Upgrade instructions**

**Review instructions**

- [ ] Userland should not need to operate on cs ids
- [ ] Events must operate on cs ids
- [ ] getSubgraph arguments
    - [ ] WorkspaceName  → internally resolved to cs id and will do lookup instantly (CACHE)
    - [ ] DimensionSpacePoint
    - [ ] VisibilityConstraints
- [ ] NodeFactory::mapNodeRowToNode  must accept WorkspaceName and ContentStreamId from outside.
    - [x] must accept ContentStreamId as parameter see https://github.com/neos/neos-development-collection/pull/4941
    - [ ] must accept WorkspaceName as parameter
- [ ] => Extend SubgraphIdentity
    - [ ] add workspaceName
    - [ ] add @internal annotation
       - [ ] and to its getters like `getIdentity` https://github.com/neos/neos-development-collection/pull/4678
- [ ] command handlers should still operate on cs ids via getSubgraphByIdentity - INTERNAL
    - [ ] by that both apis always pass the WorkspaceName
- [ ] Node should know its
    - [ ] ContentRepositoryId
    - [ ] WorkspaceName
        - [ ] NOT: ContentSubgraphIdentity.
        - [ ] NOT: ContentStreamId
    - [ ] DimensionSpacePoint it was fetched from
    - [ ] VisibilityConstraints
       - [ ] so it can be passed as single instance and eel helpers can fetch parents e.t.c in the same context 
    - [ ] OriginDimensionSpacePoint (as before)
- [ ] WorkspaceName::fromString() must be stricter without colons (so that we _could_ allow in the future for extension, e.g. WorkspaceName::detached(ContentStreamid $csId) (with an internal representation like “cs:<cs-id>”)

https://github.com/neos/neos-development-collection/pull/4943#discussion_r1548415771 but maybe manual `refresh` instead of implicit through handle.

**TODOs** as discussed with @mhsdesign and @kitsunet today:

- [ ] add workspaceName <-> contentStreamId table to content graph projection
- [ ] add workspaceName -> contentStreamId lookup to `ContentGraph::getSubgraph()` (with runtime cache in the `ContentGraph` – the same workspace will resolve to the same CS ID in one request, maybe reset runtime cache via `markStale()`)
- [ ] rewrite `ContentGraph::find*` queries to use workspaceName instead of contentStreamId (IMO it's totally fine to join the workspaceName <-> cs id table here for those queries. Alternatively we could add a dedicated lookup + runtime cache)
- [ ] remove SubgraphIdentity

Afterwards we have to find out whether there are places that need to be able to override the worspace name -> cs id mapping for tests or CR constraints. If so we could either:
* Add a separate, internal, `ContentGraph::getSubgraphInternal()` – but that would be dangerous because it is part of the public interface (even if marked internal)
* Introduce a `ContentsubgraphFactoryInterface` that can be used via CR service
* Provide a way to hook into the runtime cache, e.g. `$contentGraph->dangerousSetContentStreamIdForWorkspaceNameFooBar(...)`
* Extract workspaceName<->cs id mapping to external service that can be replaced/hooked into

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\ContentRepository\Core\SharedModel\Node;
+
+use Neos\ContentRepository\Core\DimensionSpace\DimensionSpacePoint;
+use Neos\ContentRepository\Core\Factory\ContentRepositoryId;
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php]: that seems a bit illegal that the shared kernel accesses `Factory` ? or not? Idk how we handle the namespace stuff

File: Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ *      );
+ *      $node = $subgraph->findNodeById($nodeIdentity->nodeAggregateId);
+ *
+ * @api
+ */
+final readonly class NodeIdentity implements \JsonSerializable
+{
+    public function __construct(
+        public ContentRepositoryId $contentRepositoryId,
+        public ContentStreamId $contentStreamId,
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php]: _or_ as we mostly use the workspace name for serialisation maybe that should be used here?
than we can use this dto for  https://github.com/neos/neos-development-collection/pull/4734

Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php]: Mmm, good question, I guess both have their merit, because eg. in a future conflict resolution  thing I could see need to address a contentStream that is "parked" to the side because it had conflicts and is not currently exposed via a workspace? IDK though, we could also just decide that whenever we need access to such a content stream it gets a workspace assigned?

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php]: or we make this a union type? or make two node identities? (which feels weird)

File: Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php (hunk 1 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php]: `ContentRepositoryId` should not be in that namespace o.O

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php [hunk 3]
```diff
@@ -34,6 +35,8 @@
  * call findChildNodes() {@see ContentSubgraphInterface::findChildNodes()}
  * on the subgraph.
  *
+ * The identity of a node is summarized here {@see NodeIdentity}
+ *
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Node.php:39]: Not suggesting that we change it with this PR, but it seems there is quite a lot of  overlap between `ContentSubgraphIdentity` and `NodeIdentity`

File: Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ * one can build a subgraph with the right perspective to find this node:
+ *
+ *      $workspace = $contentRepository->getWorkspaceFinder()->findOneByName($identity->workspaceName);
+ *      $subgraph = $contentGraph->getSubgraph(
+ *          $workspace->currentContentStreamId,
+ *          $nodeIdentity->dimensionSpacePoint,
+ *          // resolve also all disabled nodes
+ *          VisibilityConstraints::withoutRestrictions()
+ *      );
+ *      $node = $subgraph->findNodeById($nodeIdentity->nodeAggregateId);
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php]: This seems too cumbersome still and IMO folks should not use the `WorkspaceFinder` for things other than finding workspaces (in Neos world people can use the `NodeSerializer` instead, but...).
Maybe we can add some `ContentRepository::findNodeByIdentity(NodeIdentity $nodeIdentity)` or so in a separate PR

File: Neos.ContentRepositoryRegistry/Classes/NodeSerializer.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        private ContentRepositoryRegistry $contentRepositoryRegistry
+    ) {
+    }
+
+    public function denormalizeNodeFromIdentity(NodeIdentity $identity): Node
+    {
+        $contentRepository = $this->contentRepositoryRegistry->get($identity->contentRepositoryId);
+        $workspace = $contentRepository->getWorkspaceFinder()->findOneByName($identity->workspaceName);
+        if (!$workspace) {
+            throw new \RuntimeException(sprintf('Workspace could not be found while deserializing node NodeIdentity<%s>.', json_encode($identity, JSON_PARTIAL_OUTPUT_ON_ERROR)), 1707757488);
```
Comment (bwaidelich) [inline Neos.ContentRepositoryRegistry/Classes/NodeSerializer.php]: We mix the terms "(de)serialize" and "(de)normalize" but actually neither are correct.. In its current form it's rather a converter mixed with a repository..

The name `NodeSerializer` came from the idea to represent nodes in URLs etc.

**Review by bwaidelich — COMMENTED** [review]
Some initial comments.

My main concern (not with this PR but in general regarding this topic) is the implicit distinction between the "content stream" and the "workspace" view.

E.g. `ContentSubgraphIdentity` vs `NodeIdentity`

File: Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php]: Little `F6` action https://github.com/neos/neos-development-collection/pull/4891

File: Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php (hunk 4 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php]: yeah would prefer another pr as cluttering the `ContentRepository` must probably well discussed :D
Or rather its impossible, as we would no know which content graph implementation to use. There _could_ be multiple ones.

File: Neos.ContentRepositoryRegistry/Classes/NodeSerializer.php (hunk 5 above)
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/NodeSerializer.php]: so

```php
public function findNodeByIdentity(NodeIdentity $identity): Node;

public function convertNodeToIdentity(Node $node): NodeIdentity;
```

?

File: Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php (hunk 4 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php]: No i was wrong, via `$contentRepository->getContentGraph()` we can access the content graph.
The thing we should be careful about is not to spread too many utility methods around.

Like we already have: `\Neos\ContentRepositoryRegistry\ContentRepositoryRegistry::subgraphForNode`
And having `ContentRepository::findNodeByIdentity` would mean, that the cr has already to be fetched first by looking into the nodeIdentity which kindof defeats the purpose?

Im not sure what the best api for standalone usage is, but im glad that Neos.Neos can make it a bit easier :D

File: Neos.ContentRepositoryRegistry/Classes/NodeSerializer.php (hunk 5 above)
Comment (bwaidelich) [reply Neos.ContentRepositoryRegistry/Classes/NodeSerializer.php]: It's more about the `NodeSerializer` (and I'm not stumbling upon the naming alone, but it's usually a sign of a flawed abstraction if that doesn't fit).
Let's discuss it in our weekly later

File: Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        return new self(
+            ContentRepositoryId::fromString($array['contentRepositoryId']),
+            WorkspaceName::fromString($array['workspaceName']),
+            DimensionSpacePoint::fromArray($array['dimensionSpacePoint']),
+            NodeAggregateId::fromString($array['nodeAggregateId'])
+        );
+    }
+
+    public static function fromJsonString(string $jsonString): self
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php:60]: Nitpick: JSON is a string by definition so I would suggest to change this to:
```suggestion
    public static function fromJson(string $jsonString): self
```

..or otherwise rename `toJson()` to `toJsonString()` too

**Review by bwaidelich — COMMENTED** [review]
As recap from the weekly:
I like the `NodeIdentity` DTO and I think that it really makes sense.

`NodeSerializer` does not really make sense to me yet and I think it highlights some inconsistencies we currently have.

I could imagine that we could change the `Node` read model from:
```
subgraphIdentity:
  contentRepositoryId
  contentStreamId
  dimensionSpacePoint
  visibilityConstraints
nodeAggregateId
originDimensionSpacePoint
...
```
to
```
identity
  contentRepositoryId
  workspaceName
  dimensionSpacePoint
  nodeAggregateId
originDimensionSpacePoint
...
```

But I'm not sure about all implications yet

File: Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php (hunk 6 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php:60]: that naming scheme was not introduce by me but inferred by the codebase ;) See the dsp have the same methods i think

Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php:60]: Ah OK, dang

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php:60]: but i see your point :D Lemme think about it ^^

Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/SharedModel/Node/NodeIdentity.php:60]: It's not that important, I just stumbled upon it

**Comment by kitsunet** [comment]
@bwaidelich the suggested properties for node would prevent us from ever addressing a node that is in a contentstream without workspace attachment. I think it's fine to not plan for this from outside, but we might need to load one eg. for conflict resolution?

**Comment by bwaidelich** [comment]
I just discussed this topic with @skurfuerst and he also agrees to getting rid of the CS id in the public APIs as much as possible.

We _do_ probably still need `visibilityConstraints` in the Node read model because we use them to "stay in the same visibility context" when passing nodes via Fusion.
Personally I would prefer some required Fusion runtime variable for that (similar to `request`) but that would probably be too big of a change.
`visibilityConstraints` should not be part of the `NodeIdentity` though but it could be a property of the `Node` itself (and `SubgraphIdentity` can probably removed afterwards).

@kitsunet re

> the suggested properties for node would prevent us from ever addressing a node that is in a contentstream without workspace attachment

That's true and might be a potential issue.
Sebastian suggested to introduce something like a `VirtualWorkspaceName` (or `TransientWorkspaceName` or some similar name) that reflects a CS id without a corresponding workspace (a bit similar to a detached head in Git)
This could be a concept of the read model exclusively, i.e. 

```php
class NodeIdentity
{
    private function __construct(
        // ...
        public WorkspaceName|VirtualWorkspaceName $workspaceName
// ...
```
For write operations, i.e. commands, we would always require a `WorkspaceName`

**Comment by mhsdesign** [comment]
I implemented some wip stuff, edit moved to https://github.com/mhsdesign/neos-development-collection/tree/feature/NodeIdentityDto-original-draft

<details><summary>Details</summary>
<p>


ala (my goal was to not have to keep track of the workspaces in the content graph projection)

```php
final class ContentRepository
{
    public function getSubgraph(NodeIdentity $nodeIdentity, VisibilityConstraints $visibilityConstraints): ContentSubgraphInterface
    {
        if ($nodeIdentity->workspaceName instanceof DetachedWorkspaceName) {
            return $this->getContentGraph()->getSubgraph(
                $nodeIdentity->workspaceName->contentStreamId,
                $nodeIdentity->dimensionSpacePoint,
                $visibilityConstraints
            );
        }
        $workspace = $this->getWorkspaceFinder()->findOneByName($nodeIdentity->workspaceName);
        // ...


interface ContentGraphInterface extends ProjectionStateInterface
{
    public function getSubgraph(
        ContentStreamId|Workspace $contentStreamReference,
        DimensionSpacePoint $dimensionSpacePoint,
        VisibilityConstraints $visibilityConstraints
    ): ContentSubgraphInterface;


final readonly class NodeIdentity implements \JsonSerializable
{
    private function __construct(
        public ContentRepositoryId $contentRepositoryId,
        public WorkspaceName|DetachedWorkspaceName $workspaceName,
        public DimensionSpacePoint $dimensionSpacePoint,
        public NodeAggregateId $nodeAggregateId,
    ) {
    }


final class ContentSubgraph implements ContentSubgraphInterface
{
    // todo what if there was already a subgraph with runtime cache constructed and different / no workspace name?
    public function __construct(
        private readonly ContentRepositoryId $contentRepositoryId,
        private readonly ContentStreamId $contentStreamId,
        private readonly ?WorkspaceName $workspaceName,


final class NodeFactory
{
    public function __construct(
        private readonly ContentRepositoryId $contentRepositoryId,
        private readonly NodeTypeManager $nodeTypeManager,
        private readonly PropertyConverter $propertyConverter
    ) {
    }

    /**
     * @param array<string,string> $nodeRow Node Row from projection (<prefix>_node table)
     */
    public function mapNodeRowToNode(
        array $nodeRow,
        DimensionSpacePoint $dimensionSpacePoint,
        VisibilityConstraints $visibilityConstraints,
        ?WorkspaceName $workspaceName,
    ): Node {
        $nodeType = $this->nodeTypeManager->hasNodeType($nodeRow['nodetypename'])
            ? $this->nodeTypeManager->getNodeType($nodeRow['nodetypename'])
            : null;

        return Node::create(
            NodeIdentity::create(
                $this->contentRepositoryId,
                $workspaceName ?? DetachedWorkspaceName::fromContentStreamId(
                    ContentStreamId::fromString($nodeRow['contentstreamid'])
                ),
                $dimensionSpacePoint,
                NodeAggregateId::fromString($nodeRow['nodeaggregateid'])
            ),
```



</p>
</details>

**Comment by kitsunet** [comment]
The VirtualWorkspaceName idea seems fine, I think that should cover most bases, I like it :)

Re the visibility constraints, IMHO it makes sense to know where exactly that node is from, to refer to other nodes, eg. if I get one of those nodes handed in userland code without them being available on the node I would need to pass that in parallel to not suddenly drop into a different visibility if I need a subtree for this node.

**Comment by bwaidelich** [comment]
> eg. if I get one of those nodes handed in userland code without them being available on the node I would need to pass that in parallel

IMO that would be the "better" way to solve this: To traverse the graph from a given node, you'd need the visibility constraints, too.
Merging those into the node read model for convenience blurs the responsibilities (e.g. whether they should be part of some caching identifier depends on the scenario).

But the usecase you describe is exactly why we would need it for now because the alternative would be to introduce the visibility constraints as some kind of context to the fusion runtime

**Comment by bwaidelich** [comment]
Any update on this one?
It's one of the major outstandig changes AFAIR and it blocks some security related features.
Let me know if I can help!

**Comment by mhsdesign** [comment]
Jup. https://github.com/neos/neos-development-collection/pull/4790 should be merged first as i was about to adjust the node factory and was like wait a second, dont make it harder with the conflicts ;)

**Comment by bwaidelich** [comment]
@mhsdesign Thanks for the update and let me know if I can help with this one!

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php [hunk 7]
```diff
@@ -175,6 +196,7 @@ public function mapNodeRowsToNodeAggregate(
                 // ... so we handle occupation exactly once ...
                 $nodesByOccupiedDimensionSpacePoints[$occupiedDimensionSpacePoint->hash] = $this->mapNodeRowToNode(
                     $nodeRow,
+                    WorkspaceName::forLive(), // todo use workspace name in content graph api
```
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php:199]: `NodeAggregates` also need to operate on workspace names, but for that we would have to adjust the content graph api. That might go in a separate pr.

```
public function findNodeAggregateById(
    WorkspaceName $workspaceName,
    NodeAggregateId $nodeAggregateId
): ?NodeAggregate;
```

the cs id would be looked up internally  via the `workspaces` table.
And the cache would be similar flushed on `markStale` as the cached `subgraphs`

File: Neos.ContentRepository.Core/Classes/Feature/DimensionSpaceAdjustment/DimensionSpaceCommandHandler.php [hunk 8]
```diff
@@ -71,6 +71,8 @@ private function handleMoveDimensionSpacePoint(
         $streamName = ContentStreamEventStreamName::fromContentStreamId($command->contentStreamId)
             ->getEventStreamName();
 
+        // todo use WorkspaceName here as well but the command doesnt expose it
+        // https://github.com/neos/neos-development-collection/issues/4942
         self::requireDimensionSpacePointToBeEmptyInContentStream(
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/DimensionSpaceAdjustment/DimensionSpaceCommandHandler.php:76]: Would be great to find out why https://github.com/neos/neos-development-collection/issues/4942 is how it is.
These were the only commands untouched by bernhards refactoring.

File: Neos.ContentRepository.Core/Classes/Feature/NodeMove/NodeMove.php [hunk 9]
```diff
@@ -149,7 +150,8 @@ private function handleMoveNodeAggregate(
             $originNodeMoveMappings[] = new OriginNodeMoveMapping(
                 $movedNodeOrigin,
                 $this->resolveCoverageNodeMoveMappings(
-                    $contentStreamId,
+                    // todo UNSAFE as override via ContentStreamIdOverride will NOT be taken into account!!!
+                    $command->workspaceName,
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeMove/NodeMove.php:154]: this MUST not be used as a custom repoint via `ContentStreamIdOverride` would be ignored.
We  possibly have to pass both as touple and create a `new ContentGraph` by hand.

File: Neos.Neos/Classes/AssetUsage/Service/AssetUsageSyncService.php [hunk 10]
```diff
@@ -56,6 +56,7 @@ public function isAssetUsageStillValid(AssetUsage $usage): bool
         $dimensionSpacePoint = $usage->originDimensionSpacePoint->toDimensionSpacePoint();
 
         $subGraph = $this->contentGraph->getSubgraph(
+            // todo use workspace name instead!!!
             $usage->contentStreamId,
```
Comment (mhsdesign) [inline Neos.Neos/Classes/AssetUsage/Service/AssetUsageSyncService.php:60]: i guess adding the `workspace` to the `AssetUsage` will not be trivial, so we should just use the workspace finder beforehand and get a workspace that will be internally resolved again to a content stream?

File: Neos.ContentRepository.Core/Classes/Feature/NodeTypeChange/NodeTypeChange.php [hunk 11]
```diff
@@ -193,7 +193,7 @@ private function handleChangeNodeAggregateType(
                 $tetheredNodeName = NodeName::fromString($serializedTetheredNodeName);
 
                 $subgraph = $contentRepository->getContentGraph()->getSubgraph(
-                    $node->subgraphIdentity->contentStreamId,
+                    $node->identity->workspaceName,
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeTypeChange/NodeTypeChange.php:196]: this is illegal as well because of `ContentStreamIdOverride`.

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php (hunk 7 above)
Comment (mhsdesign) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/NodeFactory.php:199]: The problem with this would be that the write site heavily relies on the `ContentGraph` and must work on a specific contentstream id (see `ContentStreamIdOverride`)

unless we create a custom `ContentGraph` which will use our content stream, or by actually hacky checking against the state in `ContentStreamIdOverride` we wont get around that.

**Comment by mhsdesign** [comment]
Thanks for all the discussion here, but this branch is super outdated. I cherry picked all the changes to a new pr and we can continue on a blank page: https://github.com/neos/neos-development-collection/pull/5042
